File: packages/i18nify-js/src/modules/banking/__tests__/getMandateFrequencyLabel.test.ts
import getMandateFrequencyLabel from '../getMandateFrequencyLabel';

describe('mandate - getMandateFrequencyLabel', () => {
  describe('standard frequency codes', () => {
    it('should return the correct label for DAILY', () => {
      const result = getMandateFrequencyLabel('DAILY');
      expect(result.label).toBe('Daily');
      expect(result.days).toBe(1);
    });

    it('should return the correct label for WEEKLY', () => {
      const result = getMandateFrequencyLabel('WEEKLY');
      expect(result.label).toBe('Weekly');
      expect(result.days).toBe(7);
    });

    it('should return the correct label for MONTHLY', () => {
      const result = getMandateFrequencyLabel('MONTHLY');
      expect(result.label).toBe('Monthly');
      expect(result.days).toBe(30);
    });

    it('should return the correct label for QUARTERLY', () => {
      const result = getMandateFrequencyLabel('QUARTERLY');
      expect(result.label).toBe('Quarterly');
      expect(result.days).toBe(90);
    });

    it('should return the correct label for YEARLY', () => {
      const result = getMandateFrequencyLabel('YEARLY');
      expect(result.label).toBe('Yearly');
      expect(result.days).toBe(365);
    });

    it('should return undefined days for AS_PRESENTED', () => {
      const result = getMandateFrequencyLabel('AS_PRESENTED');
      expect(result.label).toBe('As Presented');
      expect(result.days).toBeUndefined();
    });
  });

  describe('alias codes', () => {
    it('BI_WEEKLY should be an alias for FORTNIGHTLY (14 days)', () => {
      const biWeekly = getMandateFrequencyLabel('BI_WEEKLY');
      const fortnightly = getMandateFrequencyLabel('FORTNIGHTLY');
      expect(biWeekly.days).toBe(14);
      expect(fortnightly.days).toBe(14);
    });

    it('SEMI_ANNUAL should be an alias for HALF_YEARLY (180 days)', () => {
      const semiAnnual = getMandateFrequencyLabel('SEMI_ANNUAL');
      const halfYearly = getMandateFrequencyLabel('HALF_YEARLY');
      expect(semiAnnual.days).toBe(180);
      expect(halfYearly.days).toBe(180);
    });

    it('ANNUAL should be an alias for YEARLY (365 days)', () => {
      expect(getMandateFrequencyLabel('ANNUAL').days).toBe(365);
    });
  });

  describe('case insensitivity', () => {
    it('should accept lowercase frequency codes', () => {
      expect(getMandateFrequencyLabel('monthly').label).toBe('Monthly');
    });

    it('should accept mixed-case frequency codes', () => {
      expect(getMandateFrequencyLabel('Weekly').label).toBe('Weekly');
    });
  });

  describe('invalid arguments', () => {
    it('should throw for an empty string', () => {
      expect(() => getMandateFrequencyLabel('')).toThrow(/frequencyCode/);
    });

    it('should throw for an unsupported frequency code', () => {
      expect(() => getMandateFrequencyLabel('HOURLY')).toThrow(/not supported/);
    });
  });
});

File: packages/i18nify-go/modules/banking/get_mandate_frequency_label_test.go
package banking

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestGetMandateFrequencyLabel(t *testing.T) {
	tests := []struct {
		name      string
		code      string
		wantLabel string
		wantDays  int
		noDays    bool // true when Days field is expected to be zero (AS_PRESENTED)
		wantErr   bool
	}{
		// ── standard codes ────────────────────────────────────────────────
		{name: "DAILY", code: "DAILY", wantLabel: "Daily", wantDays: 1},
		{name: "WEEKLY", code: "WEEKLY", wantLabel: "Weekly", wantDays: 7},
		{name: "FORTNIGHTLY", code: "FORTNIGHTLY", wantLabel: "Fortnightly", wantDays: 14},
		{name: "MONTHLY", code: "MONTHLY", wantLabel: "Monthly", wantDays: 30},
		{name: "BI_MONTHLY", code: "BI_MONTHLY", wantLabel: "Bi-Monthly", wantDays: 60},
		{name: "QUARTERLY", code: "QUARTERLY", wantLabel: "Quarterly", wantDays: 90},
		{name: "HALF_YEARLY", code: "HALF_YEARLY", wantLabel: "Half-Yearly", wantDays: 180},
		{name: "YEARLY", code: "YEARLY", wantLabel: "Yearly", wantDays: 365},
		{
			name:      "AS_PRESENTED has no days",
			code:      "AS_PRESENTED",
			wantLabel: "As Presented",
			noDays:    true,
		},

		// ── alias codes ───────────────────────────────────────────────────
		{
			name:      "BI_WEEKLY is alias for FORTNIGHTLY",
			code:      "BI_WEEKLY",
			wantLabel: "Bi-Weekly",
			wantDays:  14,
		},
		{
			name:      "SEMI_ANNUAL is alias for HALF_YEARLY",
			code:      "SEMI_ANNUAL",
			wantLabel: "Semi-Annual",
			wantDays:  180,
		},
		{
			name:      "ANNUAL is alias for YEARLY",
			code:      "ANNUAL",
			wantLabel: "Annual",
			wantDays:  365,
		},

		// ── case insensitivity ─────────────────────────────────────────────
		{name: "lowercase monthly", code: "monthly", wantLabel: "Monthly", wantDays: 30},
		{name: "mixed-case Weekly", code: "Weekly", wantLabel: "Weekly", wantDays: 7},

		// ── error cases ───────────────────────────────────────────────────
		{name: "empty string returns error", code: "", wantErr: true},
		{name: "unsupported code returns error", code: "HOURLY", wantErr: true},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := GetMandateFrequencyLabel(tt.code)
			if tt.wantErr {
				assert.Error(t, err)
				return
			}
			assert.NoError(t, err)
			assert.Equal(t, tt.wantLabel, got.Label)
			if tt.noDays {
				assert.Zero(t, got.Days)
			} else {
				assert.Equal(t, tt.wantDays, got.Days)
			}
		})
	}
}
